beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,226 @@
+//! allows the CL to load the blobs quickly from the EL's blob pool.
+//!
+//! Once the node fetches the blobs from EL, it then publishes the remaining blobs that hasn't seen
+//! on P2P gossip to the network. From PeerDAS onwards, together with the increase in blob count,
+//! broadcasting blobs requires a much higher bandwidth, and is only done by high capacity
+//! supernodes.
+use crate::observed_data_sidecars::ObservableDataSidecar;
+use crate::{BeaconChain, BeaconChainError, BeaconChainTypes, BlockError, ExecutionPayloadError};
+use itertools::Either;
+use slog::{debug, error, warn};
+use state_processing::per_block_processing::deneb::kzg_commitment_to_versioned_hash;
+use std::sync::Arc;
+use types::blob_sidecar::FixedBlobSidecarList;
+use types::{
+    BlobSidecar, DataColumnSidecar, DataColumnSidecarVec, EthSpec, FullPayload, Hash256,
+    SignedBeaconBlock,
+};
+
+pub enum BlobsOrDataColumns<E: EthSpec> {
+    Blobs(Vec<Arc<BlobSidecar<E>>>),
+    DataColumns(DataColumnSidecarVec<E>),
+}
+
+/// Fetches blobs from the EL mempool and processes them. It also broadcasts unseen blobs or
+/// data columns (PeerDAS onwards) to the network, using the supplied `publish_fn`.
+pub async fn fetch_and_process_engine_blobs<T: BeaconChainTypes>(
+    chain: Arc<BeaconChain<T>>,
+    block_root: Hash256,
+    block: Arc<SignedBeaconBlock<T::EthSpec, FullPayload<T::EthSpec>>>,
+    publish_fn: impl FnOnce(BlobsOrDataColumns<T::EthSpec>) + Send + 'static,
+) -> Result<(), BlockError<T::EthSpec>> {
+    let versioned_hashes =
+        if let Ok(kzg_commitments) = block.message().body().blob_kzg_commitments() {
+            kzg_commitments
+                .iter()
+                .map(kzg_commitment_to_versioned_hash)
+                .collect()
+        } else {
+            vec![]
+        };
+    let num_blobs = versioned_hashes.len();
+
+    if versioned_hashes.is_empty() {
+        debug!(chain.log, "Blobs from EL - none required");
+        return Ok(());
+    }
+
+    let execution_layer = chain
+        .execution_layer
+        .as_ref()
+        .ok_or(BeaconChainError::ExecutionLayerMissing)?;
+
+    debug!(
+        chain.log,
+        "Blobs from EL - start request";
+        "num_blobs" => num_blobs,
+    );
+    let response = execution_layer
+        .get_blobs(versioned_hashes)
+        .await
+        .map_err(|e| BlockError::ExecutionPayloadError(ExecutionPayloadError::RequestFailed(e)))?;
+    let num_fetched_blobs = response.iter().filter(|b| b.is_some()).count();
+    let mut all_blobs_fetched = false;
+    if num_fetched_blobs == 0 {
+        debug!(chain.log, "Blobs from EL - response with none");
+        return Ok(());
+    } else if num_fetched_blobs < num_blobs {
+        debug!(
+            chain.log,
+            "Blobs from EL - response with some";
+            "fetched" => num_fetched_blobs,
+            "total" => num_blobs,
+        );
```

Comment:
I feel this log should be moved below on the if

```
if peer_das_enabled && all_blobs_fetched {
  .. cache hit
} else if peer_das_enabled && !all_blobs_fetched {
  .. cache miss
```

Reply:
Should add in this PR a metric to track cache miss / hit from EL when retrieving blobs

Reply:
Moved logs and added metrics below